src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestTPFatTPFServerWithTranslation.java

```diff
@@ -0,0 +1,149 @@
+		
+		final TriplePatternRequest newReq = new TriplePatternRequestImpl(tp);
+		final ExecOpRequestTPFatTPFServer op = new ExecOpRequestTPFatTPFServer(newReq, fm);
+		final MaterializingIntermediateResultElementSinkWithTranslation sink = new MaterializingIntermediateResultElementSinkWithTranslation(fm.getVocabularyMapping());
+		
+		TPFResponse currentPage = null;
+		while ( currentPage == null || ! op.isLastPage(currentPage) ) {
+			// create the request for the next page (which is the first page if currentPage is null)
+			final TPFRequest pageRequest = op.createPageRequest(currentPage);
+
+			// perform the page request
+			try {
+				currentPage = op.performPageRequest( pageRequest, fedAccessMgr );
+			}
+			catch ( final FederationAccessException e ) {
+				throw new ExecOpExecutionException("Issuing a page request caused an exception: " + e.toString(), this);
+			}
+			
+			// consume the matching triples retrieved via the page request
+			op.consumeMatchingTriples( currentPage.getPayload(), sink );
+
+		}	
+	
+		return sink.getMaterializedIntermediateResult().iterator();
+
+	}
+	
+	protected Iterator<SolutionMapping> handleUnionPattern(final SPARQLUnionPattern up, final FederationAccessManager fedAccessMgr) throws ExecOpExecutionException {
+		Iterator<SPARQLGraphPattern> i = up.getSubPatterns().iterator();
+		Iterator<SolutionMapping> unionTranslation = handlePattern(i.next(), fedAccessMgr);
+		while(i.hasNext()){
+			unionTranslation = new UnionIteratorForSolMaps(unionTranslation, handlePattern(i.next(), fedAccessMgr));
+		}
+		return unionTranslation;
+	}
+	
+	protected Iterator<SolutionMapping> handleGroupPattern(final SPARQLGroupPattern gp, final FederationAccessManager fedAccessMgr) throws ExecOpExecutionException {
+		Iterator<SPARQLGraphPattern> i = gp.getSubPatterns().iterator();
+		Iterator<SolutionMapping> groupTranslation = handlePattern(i.next(), fedAccessMgr);
+		while(i.hasNext()){
+			groupTranslation = new JoiningIteratorForSolMaps(getIterableFromIterator(groupTranslation), getIterableFromIterator(handlePattern(i.next(), fedAccessMgr)));
+		}
+		return groupTranslation;
+	}
+	
+	protected Iterator<SolutionMapping> handleBGP(final BGP bgp, final FederationAccessManager fedAccessMgr) throws ExecOpExecutionException {
+		Iterator<SolutionMapping> bgpTranslation = null;
+		for(final TriplePattern i : bgp.getTriplePatterns()) {
+			if (bgpTranslation == null) {
+				bgpTranslation = handleTriplePattern(i, fedAccessMgr);
+			} else {
+				bgpTranslation = new JoiningIteratorForSolMaps(getIterableFromIterator(bgpTranslation), getIterableFromIterator(handleTriplePattern(i, fedAccessMgr)));
+			}
+		}
+		return bgpTranslation;
+	}
+
+	
+	//Source: https://www.geeksforgeeks.org/convert-iterator-to-iterable-in-java/
+	// Function to get the Spliterator
+    public static <T> Iterable<T>
+    getIterableFromIterator(Iterator<T> iterator)
```

Comment:
I'm not sure why we need this. I'm sure there is a version of the join and union iterators that Olaf made that receive iterators directly.

Reply:
the union iterator does handle either iterator or iterable inputs, but the join version requires iterables. I could make a constructor for JoiningIteratorForSolMaps which takes iterators? Or is it better to keep it this way?

Reply:
Hi, I just saw that problem and added a new constructor. We may need to change this later, but on the meantime, please use the new code.